sycl/doc/extensions/SYCL_pipes_lowering_to_SPIRV.rst

```diff
@@ -0,0 +1,111 @@
+different OpTypePipe corresponding to each read and write, with unidirectional
+access qualifiers annotated, or it can create fewer OpTypePipe objects, although
+read and write pipes must be distinct according to OpReadPipe and OpWritePipe
+rules.
+
+NOTE: The SPIR-V OpReadPipe and OpWritePipe instructions are non-blocking.
+
+Details SPIR-V representation in LLVM IR
+========================================
+
+Pipe built-ins are mangled in LLVM IR to make it SPIR-V friendly.
+As an example:
+ OpReadPipe                  -> __spirv_ReadPipe
+ OpWritePipe                 -> __spirv_WritePipe
+ OpCreatePipeFromPipeStorage -> __spirv_CreatePipeFromPipeStorage_{read|write}
+
+where 0 and 1 are representing read_only and write_only access qualifiers.
+
+More about SPIR-V representation in LLVM IR can be found under the link:
+.. _SPIRVRepresentationInLLVM.rst: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/master/docs/SPIRVRepresentationInLLVM.rst/
+
+Example of SYCL -> LLVM IR -> SPIR-V -> LLVM-IR code transformations
+====================================================================
+Consider following SYCL device code:
+.. code:: cpp
+  pipe<class some_pipe, int, 1>::write(42, SuccessCode);
+
+After compiling this code with clang we will be given following piece of IR for
+the write pipe function call (NOTE: for this implementation clang-known
+OpenCL 2.0 pipe types are reused):
+.. code:: cpp
+  define internal spir_func void @_ZN2cl4sycl4pipeIZ4mainE9some_pipeiLi1EE5writeEiRb(i32, i8* dereferenceable(1)) #4 align 2 {
+    //...
+    %12 = call spir_func i32 @_Z17__spirv_WritePipeIiEi8ocl_pipePT_ii(%opencl.pipe_wo_t addrspace(1)* %10, i32 addrspace(4)* %11, i32 4, i32 4) #8
+    //...
+  }
+
+with following declaration:
+.. code:: cpp
+  %12 = call spir_func i32 @_Z17__spirv_WritePipeIiEi8ocl_pipePT_ii(%opencl.pipe_wo_t addrspace(1)* %10, i32 addrspace(4)* %11, i32 4, i32 4) #8
+
+Yet it's looking heavy. But SPIR-V translator will drop all of these manglings
+just making a call of SPIR-V write pipe built-in:
+.. code:: cpp
+ 7 WritePipe 51 158 156 157 52 52
+
+As this implementation is based on SPIR-V representation of OpenCL program scope
+pipes, resulting code for translation back to LLVM IR from SPIR-V is good known
+OpenCL built-ins calls:
+.. code:: cpp
+  define internal spir_func void @_ZN2cl4sycl4pipeIZ4mainE9some_pipeiLi1EE5writeEiRb(i32, i8*) #0 {
+    //...
+    %9 = call spir_func i32 @__write_pipe_2(%opencl.pipe_wo_t addrspace(1)* %6, i8 addrspace(4)* %8, i32 4, i32 4)    //...
+  }
+
+again with write pipe declaration (but now it's built-in!):
+.. code:: cpp
+  declare spir_func i32 @__write_pipe_2(%opencl.pipe_wo_t addrspace(1)*, i8 addrspace(4)*, i32, i32) #0
+
+Global pipe creation itself is being done by SPIR-V built-in
+__spirv_CreatePipeFromPipeStorage_{read|write} which has no OpenCL
+representation and therefore stays in IR before and after SPIR-V tool-chain as:
```

Comment:
Unclear.

Reply:
I have rewrote this one, please check it.